File: src/core/src/main/c/cserver/link/LinkText.c
/* **********************************************************************
 * 
 * <copyright>
 * 
 *  BBN Technologies, a Verizon Company
 *  10 Moulton Street
 *  Cambridge, MA 02138
 *  (617) 873-8000
 * 
 *  Copyright (C) BBNT Solutions LLC. All rights reserved.
 * 
 * </copyright>
 * **********************************************************************
 * 
 * $Source: /cvs/distapps/openmap/src/cserver/link/src/LinkText.c,v $
 * $RCSfile: LinkText.c,v $
 * $Revision: 1.2 $
 * $Date: 2004/01/26 19:07:10 $
 * $Author: dietrick $
 * 
 * **********************************************************************
 */

#include <stdlib.h>
#include <stdio.h>

#include "LinkText.h"
#include "Link.h"
#include "GlobalConstants.h"
#include "Response.h"
#include "Request.h"

int WriteTextHeader(LinkSocket *linkSocket)
{
    int check = OK;
    check = check || WriteChars(linkSocket, TEXT_HEADER, lTEXT_HEADER);
    check = check || WriteInteger(linkSocket, GRAPHICTYPE_TEXT);
    return check;
}


int WriteLinkTextLatLon(LinkSocket *linkSocket, 
                        double lat, double lon,
                        char *String, char *fontAsString,
                        int justification,
                        LinkArgs *linkArgs, int is_unicode, 
                        int num_unicode_chars)
{
    int check = OK;

    /** Add font and string to properties. */
    check = SetKeyValuePairInLinkArgs(linkArgs, LPC_LINKTEXTSTRING, String, is_unicode, num_unicode_chars);
    if (check == -1)
      return -1; /* Memory allocation error */
    check = SetKeyValuePairInLinkArgs(linkArgs, LPC_LINKTEXTFONT, fontAsString, 0, 0);
    if (check == -1)
      return -1; /* Memory allocation error */

    check = check || WriteTextHeader(linkSocket);
    check = check || WriteInteger(linkSocket, RENDERTYPE_LATLON);
    check = check || WriteFloat(linkSocket,(float)lat);
    check = check || WriteFloat(linkSocket,(float)lon);
    check = check || WriteInteger(linkSocket, justification);
    check = check || WriteLinkArgs(linkSocket, linkArgs);
    return check;
    /* Returns -1 if there was a memory allocation error in WriteLinkArgs() */
}


int WriteLinkTextXY(LinkSocket *linkSocket,
                    int x, int y, 
                    char *String, char *fontAsString,
                    int justification,
                    LinkArgs *linkArgs, int is_unicode, int num_unicode_chars)
{
    int check = OK;  

    /** Add font and string to properties. */
    check = SetKeyValuePairInLinkArgs(linkArgs, LPC_LINKTEXTSTRING, String, is_unicode, num_unicode_chars);
    if (check == -1)
      return -1; /* Memory allocation error */

    check = SetKeyValuePairInLinkArgs(linkArgs, LPC_LINKTEXTFONT, fontAsString, 0, 0);
    if (check == -1)
      return -1; /* Memory allocation error */

    check = check || WriteTextHeader(linkSocket);
    check = check || WriteInteger(linkSocket, RENDERTYPE_XY);
    check = check || WriteInteger(linkSocket, x);
    check = check || WriteInteger(linkSocket, y);
    check = check || WriteInteger(linkSocket, justification);
    check = check || WriteLinkArgs(linkSocket, linkArgs);
    return check;
    /* Returns -1 if there was a memory allocation error in WriteLinkArgs() */
}


int WriteLinkTextOffset(LinkSocket *linkSocket,
                        double lat, double lon,
                        int x, int y,
                        char *String, char *fontAsString,
                        int justification,                 
                        LinkArgs *linkArgs, int is_unicode, 
                        int num_unicode_chars)
{
    int check = OK;

    /** Add font and string to properties. */
    check = SetKeyValuePairInLinkArgs(linkArgs, LPC_LINKTEXTSTRING, String, is_unicode, num_unicode_chars);
    if (check == -1)
      return -1; /* Memory allocation error */
    check = SetKeyValuePairInLinkArgs(linkArgs, LPC_LINKTEXTFONT, fontAsString, 0, 0);
    if (check == -1)
      return -1; /* Memory allocation error */
    
    check = check || WriteTextHeader(linkSocket);
    check = check || WriteInteger(linkSocket, RENDERTYPE_OFFSET);
    check = check || WriteFloat(linkSocket,(float)lat);
    check = check || WriteFloat(linkSocket,(float)lon);
    check = check || WriteInteger(linkSocket, x);
    check = check || WriteInteger(linkSocket, y);
    check = check || WriteInteger(linkSocket, justification);
    check = check || WriteLinkArgs(linkSocket, linkArgs);
    return check;
    /* Returns -1 if there was a memory allocation error in WriteLinkArgs() */
}

int BufferedWriteTextHeader(char *toBuffer)
{
    int byteswritten = 0;
    
    byteswritten += BufferedWriteChars(&toBuffer[byteswritten],
                                       TEXT_HEADER, lTEXT_HEADER);
    byteswritten += BufferedWriteInteger(&toBuffer[byteswritten], GRAPHICTYPE_TEXT);
    return byteswritten;
}


int BufferedWriteLinkTextLatLon(LinkSocket *linkSocket, 
                                double lat, double lon,
                                char *String, char *fontAsString,
                                int justification,
                                LinkArgs *linkArgs, int is_unicode,
                                int num_unicode_chars)
{
    int buffercount;
    char *buffer;
    int retval;

    /** Add font and string to properties. */
    retval = SetKeyValuePairInLinkArgs(linkArgs, LPC_LINKTEXTSTRING, String, 
                                       is_unicode, num_unicode_chars);
    if (retval == -1)
      return -1; /* Memory allocation error */
    retval = SetKeyValuePairInLinkArgs(linkArgs, LPC_LINKTEXTFONT, fontAsString, 0, 0);
    if (retval == -1)
      return -1; /* Memory allocation error */

    /*
      buffercount corresponds to data being written in bytes below.
    */
    buffercount = lTEXT_HEADER +  /* Bytes used by Header*/
        N_BYTES_PER_INTEGER +     /*Bytes used by GRAPHIC_TYPE. It is an Integer*/
        N_BYTES_PER_INTEGER +     /*Bytes used by RENDERTYPE identifier*/
        2*N_BYTES_PER_FLOAT +     /*Bytes used by lat and lon*/
        N_BYTES_PER_INTEGER +   /*Bytes used by justification*/
        LinkSizeOfLinkArgs(linkArgs);
    
    buffer = (char *)malloc(buffercount);
    if (buffer == NULL)
      return -1; /* Memory allocation error */

    buffercount=0;
    
    buffercount += BufferedWriteTextHeader(&buffer[buffercount]);
    buffercount += BufferedWriteInteger(&buffer[buffercount], RENDERTYPE_LATLON);
    buffercount += BufferedWriteFloat(&buffer[buffercount],(float)lat);
    buffercount += BufferedWriteFloat(&buffer[buffercount],(float)lon);    
    buffercount += BufferedWriteInteger(&buffer[buffercount], justification);
    buffercount += BufferedWriteLinkArgs(&buffer[buffercount], linkArgs);
    
    retval = WriteChars(linkSocket, buffer, buffercount);         
    free(buffer);
    return retval;
}


int BufferedWriteLinkTextXY(LinkSocket *linkSocket, int x, int y, 
                            char *String, char *fontAsString,
                            int justification, LinkArgs *linkArgs, 
                            int is_unicode, int num_unicode_chars)
{
    int buffercount;
    char *buffer;
    int retval;

    /** Add font and string to properties. */
    retval = SetKeyValuePairInLinkArgs(linkArgs, LPC_LINKTEXTSTRING, String, 
                                       is_unicode, num_unicode_chars);
    if (retval == -1)
      return -1; /* Memory allocation error */
    retval = SetKeyValuePairInLinkArgs(linkArgs, LPC_LINKTEXTFONT, fontAsString, 0, 0);
    if (retval == -1)
      return -1; /* Memory allocation error */

    /*
      buffercount corresponds to data being written in bytes below.
    */
    buffercount = lTEXT_HEADER +  /* Bytes used by Header*/
        N_BYTES_PER_INTEGER +     /*Bytes used by GRAPHIC_TYPE. It is an Integer*/
        N_BYTES_PER_INTEGER +     /*Bytes used by RENDERTYPE identifier*/
        3*N_BYTES_PER_INTEGER +   /*Bytes used by x,y,justification*/
        LinkSizeOfLinkArgs(linkArgs);
    
    buffer = (char *)malloc(buffercount);
    if (buffer == NULL)
      return -1; /* Memory allocation error */

    buffercount=0;
    
    buffercount += BufferedWriteTextHeader(&buffer[buffercount]);
    buffercount += BufferedWriteInteger(&buffer[buffercount], RENDERTYPE_XY);
    buffercount += BufferedWriteInteger(&buffer[buffercount], x);
    buffercount += BufferedWriteInteger(&buffer[buffercount], y);
    buffercount += BufferedWriteInteger(&buffer[buffercount], justification);
    buffercount += BufferedWriteLinkArgs(&buffer[buffercount], linkArgs);
    
    retval = WriteChars(linkSocket, buffer, buffercount);         
    free(buffer);
    return retval;
}


int BufferedWriteLinkTextOffset(LinkSocket *linkSocket,
                                double lat, double lon, int x, int y,
                                char *String, char *fontAsString,
                                int justification, LinkArgs *linkArgs,
                                int is_unicode, int num_unicode_chars)
{
    int buffercount;
    char *buffer;
    int retval;
    
    /** Add font and string to properties. */
    retval = SetKeyValuePairInLinkArgs(linkArgs, LPC_LINKTEXTSTRING, String, is_unicode, num_unicode_chars);
    if (retval == -1)
      return -1; /* Memory allocation error */
    
    retval = SetKeyValuePairInLinkArgs(linkArgs, LPC_LINKTEXTFONT, fontAsString, 0, 0);
    if (retval == -1)
      return -1; /*Memory allocation error */
    
    /*
      buffercount corresponds to data being written in bytes below.
    */
    buffercount = lTEXT_HEADER +  /* Bytes used by Header*/
        N_BYTES_PER_INTEGER +     /*Bytes used by GRAPHIC_TYPE. It is an Integer*/
        N_BYTES_PER_INTEGER +     /*Bytes used by RENDERTYPE identifier*/
        2*N_BYTES_PER_FLOAT +     /*Bytes used by lat and lon*/
        3*N_BYTES_PER_INTEGER +   /*Bytes used by x,y,justification*/
        LinkSizeOfLinkArgs(linkArgs);
    
    buffer = (char *)malloc(buffercount);
    if (buffer == NULL)
      return -1; /* Memory allocation error */

    buffercount=0;
    
    buffercount += BufferedWriteTextHeader(&buffer[buffercount]);
    buffercount += BufferedWriteInteger(&buffer[buffercount], RENDERTYPE_OFFSET);
    buffercount += BufferedWriteFloat(&buffer[buffercount],(float)lat);
    buffercount += BufferedWriteFloat(&buffer[buffercount],(float)lon);
    buffercount += BufferedWriteInteger(&buffer[buffercount], x);
    buffercount += BufferedWriteInteger(&buffer[buffercount], y);
    buffercount += BufferedWriteInteger(&buffer[buffercount], justification);
    buffercount += BufferedWriteLinkArgs(&buffer[buffercount], linkArgs);
    
    retval = WriteChars(linkSocket, buffer, buffercount);         
    free(buffer);
    return retval;
}

File: src/openmap/com/bbn/openmap/layer/link/LinkText.java
// **********************************************************************
// 
// <copyright>
// 
//  BBN Technologies, a Verizon Company
//  10 Moulton Street
//  Cambridge, MA 02138
//  (617) 873-8000
// 
//  Copyright (C) BBNT Solutions LLC. All rights reserved.
// 
// </copyright>
// **********************************************************************
// 
// $Source: /cvs/distapps/openmap/src/openmap/com/bbn/openmap/layer/link/LinkText.java,v $
// $RCSfile: LinkText.java,v $
// $Revision: 1.1.1.1 $
// $Date: 2003/02/14 21:35:48 $
// $Author: dietrick $
// 
// **********************************************************************


package com.bbn.openmap.layer.link;

import com.bbn.openmap.omGraphics.OMText;
import com.bbn.openmap.layer.util.LayerUtils;
import com.bbn.openmap.util.ColorFactory;
import com.bbn.openmap.util.Debug;
import com.bbn.openmap.LatLonPoint;

import java.awt.BasicStroke;
import java.io.DataInputStream;
import java.io.DataOutputStream;
import java.io.IOException;

/**
 */
public class LinkText implements LinkGraphicConstants, LinkPropertiesConstants {

    public static String DEFAULT_FONT = "-*-SansSerif-normal-o-normal--12-*-*-*-*-*-*";

   /**
     * Creates a text object, with Lat/Lon placement.
     * @param latPoint latitude of the string, in decimal degrees.
     * @param lonPoint longitude of the string, in decimal degrees.
     * @param stuff the string to be displayed.
     * @param font the Font description for the string.
     * @param just the justification of the string.
     * @param properties attributes for the graphic.
     * @param dos DataOutputStream
     * @throws IOException.
     */
    public static void write(float latPoint, float lonPoint,
			     String stuff, String font, int just, 
			     LinkProperties properties, 
			     DataOutputStream dos)
	throws IOException {
	
	dos.write(Link.TEXT_HEADER.getBytes());
	dos.writeInt(GRAPHICTYPE_TEXT);
	dos.writeInt(RENDERTYPE_LATLON);
	dos.writeFloat(latPoint);
	dos.writeFloat(lonPoint);
	dos.writeInt(just);

	properties.setProperty(LPC_LINKTEXTSTRING, stuff);
	properties.setProperty(LPC_LINKTEXTFONT, font);
	properties.write(dos);
    }

    /**
     * Creates a text object, with XY placement, and default SansSerif
     * font. 
     * @param x1 horizontal window pixel location of the string.
     * @param y1 vertical window pixel location of the string.
     * @param stuff the string to be displayed.
     * @param font the Font description for the string.
     * @param just the justification of the string
     * @param properties attributes for the graphic.
     * @param dos DataOutputStream
     * @throws IOException.
     */
    public static void write(int x1, int y1, String stuff, String font, int just, 
			     LinkProperties properties, DataOutputStream dos)
	throws IOException {

	dos.write(Link.TEXT_HEADER.getBytes());
	dos.writeInt(GRAPHICTYPE_TEXT);
	dos.writeInt(RENDERTYPE_XY);
	dos.writeInt(x1);
	dos.writeInt(y1);
	dos.writeInt(just);

	properties.setProperty(LPC_LINKTEXTSTRING, stuff);
	properties.setProperty(LPC_LINKTEXTFONT, font);
	properties.write(dos);
    }

    /**
     * Rendertype is RENDERTYPE_OFFSET.
     *
     * @param latPoint latitude of center of text/ellipse.
     * @param lonPoint longitude of center of text/ellipse.
     * @param offset_x1 # pixels to the right the center will be moved
     * from lonPoint.
     * @param offset_y1 # pixels down that the center will be moved
     * from latPoint.
     * @param aString the string to be displayed.
     * @param font the Font description for the string.
     * @param just the justification of the string.
     * @param properties attributes for the graphic.
     * @param dos DataOutputStream
     * @throws IOException.
     */
    public static void write(float latPoint, float lonPoint,
			     int offset_x1, int offset_y1, 
			     String stuff, String font, int just, 
			     LinkProperties properties, 
			     DataOutputStream dos)
	throws IOException {

	dos.write(Link.TEXT_HEADER.getBytes());
	dos.writeInt(GRAPHICTYPE_TEXT);
	dos.writeInt(RENDERTYPE_OFFSET);
	dos.writeFloat(latPoint);
	dos.writeFloat(lonPoint);
	dos.writeInt(offset_x1);
	dos.writeInt(offset_y1);

	properties.setProperty(LPC_LINKTEXTSTRING, stuff);
	properties.setProperty(LPC_LINKTEXTFONT, font);
	properties.write(dos);
    }

    /**
     * Write a text to the link.
     */
    public static void write(OMText text, Link link, LinkProperties props) 
	throws IOException {

	switch (text.getRenderType()) {
	case OMText.RENDERTYPE_LATLON:
	    write(text.getLat(), text.getLon(), text.getData(), 
		  OMText.fontToXFont(text.getFont()), text.getJustify(),
		  props, link.dos);
	    break;
	case OMText.RENDERTYPE_XY:
	    write(text.getX(), text.getY(), text.getData(), 
		  OMText.fontToXFont(text.getFont()), text.getJustify(),
		  props, link.dos);
	    break;
	case OMText.RENDERTYPE_OFFSET:
	    write(text.getLat(), text.getLon(), text.getX(), text.getY(),
		  text.getData(), OMText.fontToXFont(text.getFont()), 
		  text.getJustify(), props, link.dos);
	    break;
	default:
	    Debug.error("LinkText.write: text rendertype unknown.");
	}
    }

    public static OMText read(DataInputStream dis)
	throws IOException {

	OMText text = null;
	float lat = 0;
	float lon = 0;
	int x = 0;
	int y = 0; 
	int i = 0;
	int just = 0;
	int length;
	String string, font;

	int renderType = dis.readInt();
	
	switch (renderType){
	case RENDERTYPE_OFFSET:
	    lat = dis.readFloat();
	    lon = dis.readFloat();	    
	case RENDERTYPE_XY:
	    x = dis.readInt();
	    y = dis.readInt();	   
	    break;
	case RENDERTYPE_LATLON:
	default:
	    lat = dis.readFloat();
	    lon = dis.readFloat();	    
	}
	
	just = dis.readInt();
	
	LinkProperties properties = new LinkProperties(dis);

	string = properties.getProperty(LPC_LINKTEXTSTRING);
	font = properties.getProperty(LPC_LINKTEXTFONT);
	
	if (string == null) string = "";
	if (font == null) font = DEFAULT_FONT;
	
	switch (renderType){
	case RENDERTYPE_OFFSET:
	    text = new OMText(lat, lon, x, y, string, OMText.rebuildFont(font), just);
	    break;
	case RENDERTYPE_XY:
	    text = new OMText(x, y, string, OMText.rebuildFont(font), just);
	    break;
	case RENDERTYPE_LATLON:
	default:
	    text = new OMText(lat, lon, string, OMText.rebuildFont(font), just);
	}
	
	if (text != null){
	    text.setLinePaint(ColorFactory.parseColorFromProperties(
		properties, LPC_LINECOLOR,
		BLACK_COLOR_STRING, true));
	    text.setFillPaint(ColorFactory.parseColorFromProperties(
		properties, LPC_FILLCOLOR,
		CLEAR_COLOR_STRING, true));
	    text.setSelectPaint(ColorFactory.parseColorFromProperties(
		properties, LPC_HIGHLIGHTCOLOR,
		BLACK_COLOR_STRING, true));
	    text.setStroke(new BasicStroke(LayerUtils.intFromProperties(
		properties, LPC_LINEWIDTH, 1)));
	    text.setAppObject(properties);
	}

	return text;
    }


}
